crates/ark/src/lsp/folding_range.rs

```diff
@@ -0,0 +1,299 @@
+    let mut comment_stack: Vec<Vec<(usize, usize)>> = vec![Vec::new()];
+    let mut region_marker: Option<usize> = None;
+    let mut cell_marker: Option<usize> = None;
+
+    let mut line_iter = text.lines().enumerate().peekable();
+    let mut line_count = 0;
+    while let Some((line_idx, line)) = line_iter.next() {
+        line_count += 1;
+        let line_text = line.to_string();
+        (folding_ranges, bracket_stack, comment_stack) = bracket_processor(
+            folding_ranges,
+            bracket_stack,
+            comment_stack,
+            line_idx,
+            &line_text,
+        );
+        (folding_ranges, comment_stack) =
+            comment_processor(folding_ranges, comment_stack, line_idx, &line_text);
+        (folding_ranges, region_marker) =
+            region_processor(folding_ranges, region_marker, line_idx, &line_text);
+        (folding_ranges, cell_marker) =
+            cell_processor(folding_ranges, cell_marker, line_idx, &line_text);
+    }
+
+    // Use `end_bracket_handler` to close any remaining comments
+    // There should only be one element in `comment_stack` though
+    while !comment_stack.is_empty() && !comment_stack.last().unwrap().is_empty() {
+        (folding_ranges, comment_stack) =
+            end_bracket_handler(folding_ranges, comment_stack, line_count);
+    }
+    // Deal with unclosed cells
+    if cell_marker.is_some() {
+        let fold_range = comment_range(cell_marker.unwrap(), line_count - 1);
+        folding_ranges.push(fold_range);
+        cell_marker = None;
+    }
+
+    // Log the final folding ranges and comment stacks
+    // log_info!("folding_ranges: {:#?}", folding_ranges); // Contains all folding ranges
+    // log_info!("comment_stack: {:#?}", comment_stack); // Should be empty
+    // log_info!("bracket_stack: {:#?}", bracket_stack); // Should be empty
+    // log_info!("region_marker: {:#?}", region_marker); // Should be None
+    // log_info!("cell_marker: {:#?}", cell_marker); // Should be None
+
+    Ok(folding_ranges)
+}
+
+fn bracket_processor(
+    mut folding_ranges: Vec<FoldingRange>,
+    mut bracket_stack: Vec<(usize, usize)>,
+    mut comment_stack: Vec<Vec<(usize, usize)>>,
+    line_idx: usize,
+    line_text: &str,
+) -> (
+    Vec<FoldingRange>,
+    Vec<(usize, usize)>,
+    Vec<Vec<(usize, usize)>>,
+) {
+    // Remove any trailing comments (starting with #) and \n in line_text
+    let line_text = line_text.split('#').next().unwrap_or("").trim_end();
+    let mut whitespace_count = 0;
+
+    // Iterate over each character in line_text to find the positions of `{` and `}`
+    for (char_idx, c) in line_text.char_indices() {
+        match c {
+            '{' | '(' | '[' => {
+                bracket_stack.push((line_idx, char_idx));
+                comment_stack.push(Vec::new());
+            },
+            '}' | ')' | ']' => {
+                (folding_ranges, comment_stack) =
```

Comment:
Matching on characters like this will work incorrectly with delimiter characters in strings for instance.

Reply:
Ohh I see! In this case it is indeed better to do it with a syntax tree.